experimental/segmenter/src/math_helper.rs

```diff
@@ -2,59 +2,338 @@
+        assert_eq!(M, D - 1);
+        let (range, dims) = self.as_borrowed().submatrix_range(index);
+        let data = &self.data.get(range)?;
+        Some(MatrixBorrowed { data, dims })
+    }
+
+    pub fn as_mut(&mut self) -> MatrixBorrowedMut<D> {
+        MatrixBorrowedMut {
+            data: &mut self.data,
+            dims: self.dims,
+        }
+    }
+
+    #[inline]
+    pub fn submatrix_mut<const M: usize>(&mut self, index: usize) -> Option<MatrixBorrowedMut<M>> {
+        assert_eq!(M, D - 1);
+        let (range, dims) = self.as_borrowed().submatrix_range(index);
+        let data = self.data.get_mut(range)?;
+        Some(MatrixBorrowedMut { data, dims })
+    }
 }
 
-/// `softmax` gets a 1d array of `f32` numbers, and compute the softmax probability for each element in the array.
-pub fn softmax(arr: Array1<f32>) -> Array1<f32> {
-    let sm = arr.fold(0.0, |sm, v| sm + v.exp());
-    arr.map(|v| v.exp() / sm)
+#[derive(Debug, Clone, Copy)]
+pub struct MatrixBorrowed<'a, const D: usize> {
+    data: &'a [f32],
+    dims: [usize; D],
 }
 
-/// `max_arr1` returns the index of the maximum value in a 1d array.
-pub fn max_arr1(arr: ArrayBase<ViewRepr<&f32>, Dim<[usize; 1]>>) -> usize {
-    // No argmax in rust-ndarray (https://github.com/rust-ndarray/ndarray/issues/416)
-    let mut mx: f32 = 0.0;
-    let mut ind = 0;
-    arr.indexed_iter().for_each(|(i, v)| {
-        if mx < *v {
-            mx = *v;
-            ind = i
+impl<'a, const D: usize> MatrixBorrowed<'a, D> {
+    #[cfg(debug_assertions)]
+    pub fn debug_assert_dims(&self, dims: [usize; D]) {
+        debug_assert_eq!(dims, self.dims);
+        let expected_len = dims.iter().product::<usize>();
+        debug_assert_eq!(expected_len, self.data.len());
+    }
+
+    pub fn as_slice(&self) -> &'a [f32] {
+        self.data
+    }
+
+    #[allow(clippy::wrong_self_convention)] // same convention as slice::to_vec
+    pub fn to_owned(&self) -> MatrixOwned<D> {
+        MatrixOwned {
+            data: self.data.to_vec(),
+            dims: self.dims,
         }
-    });
-    ind
+    }
+
+    pub fn argmax(&self) -> usize {
+        let mut mx: f32 = 0.0;
+        let mut ind = 0;
+        self.data.iter().enumerate().for_each(|(i, v)| {
+            if mx < *v {
+                mx = *v;
+                ind = i
+            }
+        });
+        ind
+    }
+
+    #[inline]
+    pub fn submatrix<const M: usize>(&self, index: usize) -> Option<MatrixBorrowed<'a, M>> {
+        // This assertion should always fail and be elided at compile time
+        assert_eq!(M, D - 1);
+        let (range, dims) = self.submatrix_range(index);
+        let data = &self.data.get(range)?;
+        Some(MatrixBorrowed { data, dims })
+    }
+
+    #[inline]
+    fn submatrix_range<const M: usize>(&self, index: usize) -> (Range<usize>, [usize; M]) {
+        // This assertion should always fail and be elided at compile time
```

Comment:
nit: "should always fail" seems misleading. "This assertion is only based on const generics and if it is supposed to fail it will always be known at compile time and get elided" or something, perhaps?

Reply:
Improved the comment